File: src/main/java/com/workforce/hrm/service/EmployeeService.java
package com.workforce.hrm.service;

import java.time.LocalDate;
import java.util.List;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import com.workforce.hrm.dto.request.EmployeeRequestDTO;
import com.workforce.hrm.dto.response.EmployeeResponseDTO;
import com.workforce.hrm.enums.EmployeeStatus;

public interface EmployeeService {

    // =========================================================
    // CREATE EMPLOYEE
    // =========================================================

    EmployeeResponseDTO createEmployee(
            EmployeeRequestDTO request);


    // =========================================================
    // GET ALL EMPLOYEES
    // =========================================================

    List<EmployeeResponseDTO> getAllEmployees();

    Page<EmployeeResponseDTO> getEmployees(
            String search,
            EmployeeStatus status,
            Long companyId,
            Long departmentId,
            Long designationId,
            Pageable pageable);

    Page<EmployeeResponseDTO> getEmployeesForReport(
            String search,
            EmployeeStatus status,
            Long departmentId,
            Long designationId,
            LocalDate fromDate,
            LocalDate toDate,
            Pageable pageable);


    // =========================================================
    // GET EMPLOYEE BY ID
    // =========================================================

    EmployeeResponseDTO getEmployeeById(
            Long id);


    // =========================================================
    // UPDATE EMPLOYEE
    // =========================================================

    EmployeeResponseDTO updateEmployee(
            Long id,
            EmployeeRequestDTO request);

    EmployeeResponseDTO updateEmployeeStatus(
            Long id,
            EmployeeStatus status);


    // =========================================================
    // DELETE EMPLOYEE
    // =========================================================

    void deleteEmployee(
            Long id);


    // =========================================================
    // GET EMPLOYEE BY CODE
    // =========================================================

    EmployeeResponseDTO getEmployeeByCode(
            String code);


    // =========================================================
    // GET EMPLOYEES BY DEPARTMENT
    // =========================================================

    List<EmployeeResponseDTO> getEmployeesByDepartment(
            Long departmentId);


    // =========================================================
    // GET EMPLOYEES BY STATUS
    // =========================================================

    List<EmployeeResponseDTO> getEmployeesByStatus(
            EmployeeStatus status);


    // =========================================================
    // SEARCH EMPLOYEES
    // =========================================================

    Page<EmployeeResponseDTO> searchEmployees(
            String keyword,
            Pageable pageable);


    // =========================================================
    // EMPLOYEE PROFILE
    // =========================================================

    EmployeeResponseDTO getEmployeeProfile(
            Long id);
}

File: hrm-portal-frontend/src/services/employeeService.js
import api from "./api";

// ============================================================
// EMPLOYEE SERVICE
// ============================================================

const employeeService = {

  // ==========================================================
  // GET EMPLOYEES
  // ==========================================================

  async getEmployees({
    page = 0,
    size = 20,
    sortBy = "id",
    direction = "asc",
    search = "",
    status = "",
    companyId = "",
    departmentId = "",
    designationId = ""
  } = {}) {

    const params = {
      page,
      size,
      sortBy,
      direction
    };


    // --------------------------------------------------------
    // Optional parameters
    // --------------------------------------------------------

    if (search?.trim()) {
      params.search =
        search.trim();
    }


    if (status) {
      params.status =
        status;
    }


    if (companyId) {
      params.companyId =
        companyId;
    }


    if (departmentId) {
      params.departmentId =
        departmentId;
    }


    if (designationId) {
      params.designationId =
        designationId;
    }


    const response =
      await api.get(
        "/employees",
        {
          params
        }
      );


    return response.data;
  },


  // ==========================================================
  // GET EMPLOYEE BY ID
  // ==========================================================

  async getEmployeeById(
    employeeId
  ) {

    if (!employeeId) {
      throw new Error(
        "Employee ID is required."
      );
    }


    const response =
      await api.get(
        `/employees/${employeeId}`
      );


    return response.data;
  },


  // ==========================================================
  // CREATE EMPLOYEE
  // ==========================================================

  async createEmployee(
    payload
  ) {

    if (!payload) {
      throw new Error(
        "Employee data is required."
      );
    }


    const response =
      await api.post(
        "/employees",
        payload
      );


    return response.data;
  },


  // ==========================================================
  // UPDATE EMPLOYEE
  // ==========================================================

  async updateEmployee(
    employeeId,
    payload
  ) {

    if (!employeeId) {
      throw new Error(
        "Employee ID is required."
      );
    }


    if (!payload) {
      throw new Error(
        "Employee data is required."
      );
    }


    const response =
      await api.put(
        `/employees/${employeeId}`,
        payload
      );


    return response.data;
  },


  // ==========================================================
  // DELETE EMPLOYEE
  // ==========================================================

  async deleteEmployee(
    employeeId
  ) {

    if (!employeeId) {
      throw new Error(
        "Employee ID is required."
      );
    }


    const response =
      await api.delete(
        `/employees/${employeeId}`
      );


    return response.data;
  },


  // ==========================================================
  // UPDATE EMPLOYEE STATUS
  // ==========================================================

  async updateStatus(
    employeeId,
    status
  ) {

    if (!employeeId) {
      throw new Error(
        "Employee ID is required."
      );
    }


    if (!status) {
      throw new Error(
        "Employee status is required."
      );
    }


    const response =
      await api.patch(
        `/employees/${employeeId}/status`,
        {
          status
        }
      );


    return response.data;
  },


  // ==========================================================
  // ACTIVATE EMPLOYEE
  // ==========================================================

  async activateEmployee(
    employeeId
  ) {

    return this.updateStatus(
      employeeId,
      "ACTIVE"
    );
  },


  // ==========================================================
  // DEACTIVATE EMPLOYEE
  // ==========================================================

  async deactivateEmployee(
    employeeId
  ) {

    return this.updateStatus(
      employeeId,
      "INACTIVE"
    );
  }

};


export default employeeService;